### PR TITLE
[Performance][DSA-CP] Optimize local token metadata computation with fused Triton kernel and caching

### PR DESCRIPTION
**Pull request by Csrayz**
### What this PR does / why we need it?

In DSA context-parallel mode, `AscendDSACPMetadataBuilder.build()` runs once per attention group within a step. Under DeepSeek-V4's dsv4 attention, there can be 8 build() per step, and each build independently recomputes metadata from scratch.

Across all groups, the following data are identical:

- `query_start_loc`, `seq_lens`, `num_reqs`, `num_input_tokens`
- Clamp + cumsum + offset + mask results
- RoPE local slices
- `start_pos`

This PR caches these intermediate results so that only the first group computes them; subsequent groups reuse the cached values. Additionally, the fused `_build_local_metadata_npu_kernel` is introduced to reduce kernel launch overhead.

The metadata building phase dropped from approximately 23ms to 10ms per step (measured on DeepSeek-V4 Flash, mtp=1).

before (main)

c5fa490ea22b40dbaafa4c2dab1f760d_IMAGE_1784202330638

after (this pr)

7281c93267ea4205b34526f9fb6ebd3a_IMAGE_1784202321384

**Changes summary**

- New `_ensure_gpu_local_metadata` method that computes GPU local metadata once and caches the result in `common_ratio_to_sas_metadata["_gpu_local"]`.
- Cross-group caching for CPU local metadata (`"_cpu_local"`) and RoPE local slices (`"_rope_local"`).
- New Triton kernel `_build_local_metadata_npu_kernel` that fuses clamp, cumsum, offset, mask, and start_pos into a single kernel launch.
- Refactored `build_req_metadata` to use the cached results and inlined RoPE slicing.
- Cleaned up `_build_local_token_metadata` by removing unused parameters (`input_positions`, `use_cache`, `seq_lens_q`).

### Does this PR introduce *any* user-facing change?

No. Behavior is identical.

### How was this patch tested?

- Added a unit test for `RopeDataProxy.pad_to`.
- GMS8K evaluation results passed.

```
before (main)
+---------+-----------+----------+----------+-------+---------+---------+
| Model   | Dataset   | Metric   | Subset   |   Num |   Score | Cat.0   |
+=========+===========+==========+==========+=======+=========+=========+
| dsv4-f    | gsm8k     | mean_acc | main     |  1319 |  0.9682 | default |
+---------+-----------+----------+----------+-------+---------+---------+ 
```

aftet (this pr)

```
+---------+-----------+----------+----------+-------+---------+---------+
| Model   | Dataset   | Metric   | Subset   |   Num |   Score | Cat.0   |
+=========+===========+==========+==========+=======+=========+=========+
| dsv4-f    | gsm8k     | mean_acc | main     |  1319 |  0.9712 | default |
+---------+-----------+----------+----------+-------+---------+---------+ 
```

- vLLM version: v0.25.1
- vLLM main: https://github.com/vllm-project/vllm/commit/54503ecec0f3ac31e5ecfc5f28652e4cc42307b5

### CODE REVIEW

File: vllm_ascend/attention/context_parallel/dsa_cp.py [hunk 1]
```diff
@@ -733,51 +837,58 @@ def _build_local_token_metadata(
...
-        if local_query_start_loc is not None:
-            local_query_start_loc[1 : num_reqs + 1] = torch.cumsum(local_query_lens, dim=0)
-        else:
-            local_query_start_loc = torch.cat(
-                [
-                    torch.tensor([0], dtype=local_query_lens.dtype, device=local_query_lens.device),
-                    torch.cumsum(local_query_lens, dim=0),
-                ],
-                0,
+        if query_start_loc.device.type != "cpu" and HAS_TRITON and num_reqs <= 1024:
```
Comment (linfeng-yuan) [inline vllm_ascend/attention/context_parallel/dsa_cp.py]: 1. Why we should set a magic number 1024 here?
2. PTAL @ZT-AIA 
3.  Move triton implementation to triton dir.
4.  add e2e cases

**Comment by linfeng-yuan** [comment]
@pisceskkk PTAL Threre's a performance optimization for dsa_cp with dsv4.

File: vllm_ascend/attention/context_parallel/dsa_cp.py (hunk 1 above)
Comment (Csrayz) [reply vllm_ascend/attention/context_parallel/dsa_cp.py]: 1. 1024 was just a conservative upper bound; our actual batch size never reaches it, so the check is always true and removing it is safe.
2. ~
3. done
4. done

**Comment by Csrayz** [comment]
> @pisceskkk PTAL Threre's a performance optimization for dsa_cp with dsv4.

@pisceskkk, sorry to ping again. I have another fix for dsa_cp in the MTP scenario, which modifies the same function and will cause merge conflicts. Could you please take a look at the current performance optimization PR? Thanks!

File: vllm_ascend/attention/context_parallel/dsa_cp.py [hunk 2]
```diff
@@ -554,66 +551,135 @@ def _num_compressor_metadata_rows(
         num_tokens = self.num_actual_tokens
         return min(num_tokens, num_tokens // self.compressor_ratio + common_attn_metadata.num_reqs)
 
+    def _ensure_gpu_local_metadata(
```
Comment (pisceskkk) [inline vllm_ascend/attention/context_parallel/dsa_cp.py]: I think this should be renamed as `_ensure_device_local_metadata`

**Review by pisceskkk — APPROVED** [review]
Sorry for later reply.  LGTM, leave a little suggestion.

File: vllm_ascend/attention/context_parallel/dsa_cp.py (hunk 2 above)
Comment (Csrayz) [reply vllm_ascend/attention/context_parallel/dsa_cp.py]: Sure, I'll push a new commit shortly to address it.

Also, just a heads-up: I'll be submitting a few more optimizations and bug fixes for dsa_cp in the coming days. I'd really appreciate your review on those as well if you have the bandwidth. Thanks in advance!s

**Comment by Csrayz** [comment]
@kunpengW-code, linfeng-yuan is currently unreachable. This PR's CI is all green and pisceskkk has already approved. Could you take a look and approve/merge when you get a chance?

**Review by linfeng-yuan — APPROVED** [review]
